tests/41end-to-end-keys/08-cross-signing.pl

```diff
@@ -580,6 +580,110 @@
       });
    };
 
+test "uploading signed devices gets propagated over federation",
+   requires => [ local_user_fixture(), remote_user_fixture() ],
+
+   do => sub {
+      my ( $user1, $user2 ) = @_;
+
+      my $user1_id = $user1->user_id;
+      my $user1_device = $user1->device_id;
+      my $user2_id = $user2->user_id;
+      my $user2_device = $user2->device_id;
+
+      my $room_id;
+
+      my ( $master_pubkey, $master_secret_key ) = $crypto_sign->keypair( decode_base64( "2lonYOM6xYKdEsO+6KrC766xBcHnYnim1x/4LFGF8B0" ) );
+      my ( $self_signing_pubkey, $self_signing_secret_key ) = $crypto_sign->keypair( decode_base64( "HvQBbU+hc2Zr+JP1sE0XwBe1pfZZEYtJNPJLZJtS+F8" ) );
+      my $self_signing_key = {
+         # private key: HvQBbU+hc2Zr+JP1sE0XwBe1pfZZEYtJNPJLZJtS+F8
+         "user_id" => $user2_id,
+         "usage" => ["self_signing"],
+         "keys" => {
+            "ed25519:EmkqvokUn8p+vQAGZitOk4PWjp7Ukp3txV2TbMPEiBQ"
+                => "EmkqvokUn8p+vQAGZitOk4PWjp7Ukp3txV2TbMPEiBQ",
+         },
+      };
+      sign_json(
+         $self_signing_key, secret_key => $master_secret_key,
+         origin => $user2_id, key_id => "ed25519:nqOvzeuGWT/sRx3h7+MHoInYj3Uk2LD/unI9kDYcHwk"
+          );
+
+      my $device = {
+          "user_id" => $user2_id,
+          "device_id" => $user2_device,
+          "algorithms" => ["m.olm.curve25519-aes-sha256", "m.megolm.v1.aes-sha"],
+          "keys" => {
+              "curve25519:".$user2_device => "curve25519+key",
+              "ed25519:".$user2_device => "ed25519+key",
+          }
+      };
+      my $cross_signature;
+
+      matrix_put_e2e_keys( $user2, device_keys => $device)->then( sub {
+         matrix_set_cross_signing_key( $user2, {
+             "auth" => {
+                 "type"     => "m.login.password",
+                 "user"     => $user2_id,
+                 "password" => $user2->password,
+             },
+             "master_key" => {
+                 # private key: 2lonYOM6xYKdEsO+6KrC766xBcHnYnim1x/4LFGF8B0
+                 "user_id" => $user2_id,
+                 "usage" => ["master"],
+                 "keys" => {
+                     "ed25519:nqOvzeuGWT/sRx3h7+MHoInYj3Uk2LD/unI9kDYcHwk"
+                         => "nqOvzeuGWT/sRx3h7+MHoInYj3Uk2LD/unI9kDYcHwk",
+                 },
+             },
+             "self_signing_key" => $self_signing_key,
+          });
+      })->then( sub {
+         matrix_create_room( $user1 );
+      })->then( sub {
+         ( $room_id ) = @_;
+         matrix_sync( $user1 );
```

Comment:
Do we need to sync here?

Reply:
I don't know. Probably not.  I think I just cargo-culted it from somewhere else.  I'll try without and if it still works, I'll drop it.

Reply:
It seems to be necessary, for some strange reason.  If I leave it off, then synapse will return a 500 at some later sync. :fearful: The backtrace is:

```
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/venv/lib/python3.5/site-packages/synapse/http/server.py", line 78, in wrapped_request_handler
    await h(self, request)
  File "/venv/lib/python3.5/site-packages/synapse/http/server.py", line 331, in _async_render
    callback_return = await callback_return
  File "/venv/lib/python3.5/site-packages/synapse/rest/client/v2_alpha/sync.py", line 178, in on_GET
    full_state=full_state,
  File "/venv/lib/python3.5/site-packages/synapse/handlers/sync.py", line 286, in wait_for_sync_for_user
    full_state,
  File "/venv/lib/python3.5/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
  File "/venv/lib/python3.5/site-packages/synapse/handlers/sync.py", line 323, in _wait_for_sync_for_user
    from_token=since_token,
  File "/venv/lib/python3.5/site-packages/synapse/notifier.py", line 352, in wait_for_events
    listener = user_stream.new_listener(prev_token)
  File "/venv/lib/python3.5/site-packages/synapse/notifier.py", line 137, in new_listener
    if self.last_notified_token.is_after(token):
  File "/venv/lib/python3.5/site-packages/synapse/types.py", line 376, in is_after
    (other.room_stream_id < self.room_stream_id)
  File "/venv/lib/python3.5/site-packages/synapse/types.py", line 371, in room_stream_id
    return int(self.room_key[1:].split("-")[-1])
ValueError: invalid literal for int() with base 10: ''
```

which looks like fun.

Reply:
Ah, looks like `sync_until_user_in_device_list` uses `matrix_sync_again`, which sets an empty since token